### PR TITLE
add grafanaPlugins.grafana-clickhouse-datasource

### PR DESCRIPTION
**Pull request by majiru**
###### Description of changes

Adds the grafana clickhouse integration. This plugin is platform specific so some minor
refactoring of the existing fetchurl was done. I am quite new to nix so if there is a better
way of doing this refactor let me know.

[Homepage for the plugin](https://grafana.com/grafana/plugins/grafana-clickhouse-datasource/)

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yu-re-ka** [comment]
Hi! I happen to need the same thing. Your approach is fine for downloading the appropriate zipfile, however I would always prefer building things from source. In this case maybe a compromise of downloading the x86_64-linux zipfile (regardless of our platform), then building the go binary from source and replacing it in the zipfile.
Let me know if you want to try this, or if you would like me to pick this up in a separate PR.

**Review by yu-re-ka — COMMENTED** [review]
instead of adding a new `hashFn` parameter, the `zipHash` parameter could be set to an attributeset:

```
zipHash = {
  x86_64-linux = "sha256-6WNrkp5ZtQM0CxWjDsGchpIObSXI56Zmef6QlNYmvsg";
};
```

Then, access it as:

```
if lib.isAttrs zipHash then
  zipHash.${stdenvNoCC.targetPlatform.system} or throw "unsupported system";
else zipHash
```

**Comment by majiru** [comment]
Yeah I think building from source, where it makes sense for the specific plugin, here would be great. I am not sure exactly what would be the easiest way to do that would be, does it make sense to start treating the plugins as more full fledged packages instead of having these inherited defaults?

I also think your solution of making `zipHash` an attributeset is quite nice, I will try and refactor this to use that approach.

File: pkgs/servers/monitoring/grafana/plugins/grafana-plugin.nix [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
-  src = fetchurl {
-    name = "${pname}-${version}.zip";
-    url = "https://grafana.com/api/plugins/${pname}/versions/${version}/download";
-    hash = zipHash;
-  };
+  src = (fetchurl {
+    name = "${pname}-${version}-${plat}.zip";
+    hash = if lib.isAttrs zipHash then zipHash.${plat} or throw "unsupported system" else zipHash;
+    url = "https://grafana.com/api/plugins/${pname}/versions/${version}/download" +
+      (if plat == "x86_64-linux" then "?os=linux&arch=amd64"
```
Comment (yu-re-ka) [inline pkgs/servers/monitoring/grafana/plugins/grafana-plugin.nix:11]: I would use an attrset lookup with a throw fallback here as well:

```nix
{
  x86_64-linux = "?os=linux&arch=amd64";
  aarch64-linux = "?os=linux&arch=arm64";
  x86_64-darwin = "?os=darwin&arch=amd64";
  aarch64-darwin = "?os=darwin&arch=arm64";
}.${plat} or throw "unknown system"
```

**Review by yu-re-ka — APPROVED** [review]
Looks OK 🚀

**Comment by yu-re-ka** [comment]
@ofborg build grafanaPlugins.grafana-clickhouse-datasource

**Comment by majiru** [comment]
When testing these new changes locally I caught seemingly one typo, and also couldn't quite get the `attributeSet.key or throw "..."` to quite work. https://github.com/majiru/nixpkgs/commit/cfada95c5b0fb80d45e412d2940d27a1af39f38a is what I changed to get it building locally again. Didn't want to push over your code, so I pushed it to another branch.

**Comment by yu-re-ka** [comment]
I pushed a version that works. Needed to add some parenthesis.

**Comment by majiru** [comment]
Thanks! Looks all good from my end.

**Comment by yu-re-ka** [comment]
@ofborg build grafanaPlugins.grafana-clickhouse-datasource

**Comment by yu-re-ka** [comment]
There was another missing parenthesis in the second `or throw`.

**Pull request merged**